ambry-utils/src/main/java/com.github.ambry.utils/Utils.java

```diff
@@ -210,6 +211,64 @@ public static ByteBuffer readShortBuffer(DataInputStream input) throws IOExcepti
     return buffer;
   }
 
+  /**
+   * Creating a {@link ByteBufferInputStream} from the {@link CrcInputStream} by sharing the underlying memory if the
+   * crcStream is built upon a {@link ByteBufferDataInputStream}.
+   * @param crcStream The crcStream to read {@link ByteBuffer} out.
+   * @param dataSize The size of {@link ByteBuffer}.
+   * @return The {@link ByteBufferInputStream}
+   * @throws IOException Unexpected IO errors.
+   */
+  public static ByteBufferInputStream getByteBufferInputStreamFromCRCInputStream(CrcInputStream crcStream, int dataSize)
+      throws IOException {
+    ByteBufferInputStream output;
+    InputStream inputStream = crcStream.getUnderlyingInputStream();
+    if (inputStream instanceof ByteBufferDataInputStream) {
+      ByteBuffer byteBuffer = ((ByteBufferDataInputStream) inputStream).getBuffer();
+      int startIndex = byteBuffer.position();
+      int oldLimit = byteBuffer.limit();
+
+      byteBuffer.limit(startIndex + dataSize);
+      ByteBuffer dataBuffer = byteBuffer.slice();
+      crcStream.updateCrc(dataBuffer.duplicate());
+
+      output = new ByteBufferInputStream(dataBuffer);
+      byteBuffer.limit(oldLimit);
+      // Change the byte buffer's position as if the data is fetched.
+      byteBuffer.position(startIndex + dataSize);
+    } else {
+      output = new ByteBufferInputStream(crcStream, dataSize);
+    }
+    return output;
+  }
+
+  public static ByteBuffer readByteBufferFromCRCInputStream(CrcInputStream crcStream, int dataSize) throws IOException {
+    ByteBuffer result;
+    InputStream inputStream = crcStream.getUnderlyingInputStream();
```

Comment:
Lot of duplication in next 9 lines compared to previous method.  Possible to reuse?